File: fashion_predict.r
install.packages("tidyr")
library(keras3)
library(ggplot2)
library(dplyr)
library(tidyr)
library(this.path)

# Get directory of THIS script
script_dir <- this.dir()
setwd(script_dir)

class_names <- c(
  "T-shirt/top", "Trouser", "Pullover", "Dress", "Coat",
  "Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"
)

load_assets <- function() {
  list(
    model = load_model("fashion_mnist_cnn.keras"),
    x_test = readRDS("test_images.rds"),
    y_test = readRDS("test_labels.rds")
  )
}

prepare_image_data <- function(images, actual_labels, predicted_labels) {
  # Convert 4D array to list of data frames
  image_data <- lapply(1:dim(images)[1], function(i) {
    as.data.frame.table(images[i, , , 1]) %>%
      mutate(x = as.numeric(Var2),
             y = 29 - as.numeric(Var1),  # Flip y-axis for proper orientation
             value = Freq,
             label = paste0("Pred: ", class_names[predicted_labels[i] + 1], "\n",
                            "True: ", class_names[actual_labels[i] + 1])) %>%
      select(x, y, value, label)
  })
  
  bind_rows(image_data)
}

plot_predictions <- function(model, x_test, y_test, num_images = 3) {
  # Random sample
  indices <- sample(dim(x_test)[1], num_images)
  images <- x_test[indices, , , , drop = FALSE]
  actual <- y_test[indices]
  
  # Predictions
  predictions <- predict(model, images)
  predicted <- max.col(predictions) - 1
  
  # Prepare data for ggplot
  plot_data <- prepare_image_data(images, actual, predicted)
  
  # Create plot
  ggplot(plot_data, aes(x = x, y = y, fill = value)) +
    geom_tile() +
    scale_fill_gradient(low = "white", high = "black") +
    facet_wrap(~ label, nrow = 1) +
    theme_void() +
    theme(
      strip.text = element_text(size = 10, face = "bold"),
      legend.position = "none"
    )
}

main <- function() {
  assets <- load_assets()
  print(plot_predictions(assets$model, assets$x_test, assets$y_test))
}

main()

File: fashion_predict.py
import os
import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras

# Get the directory where this script is located
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

# Class names for Fashion MNIST
CLASS_NAMES = [
    'T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
    'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'
]

def load_assets():
    # Load model and test data from script directory
    model = keras.models.load_model(os.path.join(SCRIPT_DIR, 'fashion_mnist_cnn.keras'))
    x_test = np.load(os.path.join(SCRIPT_DIR, 'test_images.npy'))
    y_test = np.load(os.path.join(SCRIPT_DIR, 'test_labels.npy'))
    return model, x_test, y_test

def plot_predictions(model, x_test, y_test, num_images=3):
    # Randomly select images from test set
    random_indices = np.random.choice(len(x_test), num_images, replace=False)
    images = x_test[random_indices]
    true_labels = y_test[random_indices]
    
    # Generate predictions
    predictions = model.predict(images, verbose=0)
    predicted_labels = np.argmax(predictions, axis=1)

    # Create plot with predictions
    plt.figure(figsize=(12, 6))
    for i in range(num_images):
        plt.subplot(1, num_images, i+1)
        plt.imshow(images[i].squeeze(), cmap='gray')  # Remove channel dimension for plotting
        plt.title(f"Predicted: {CLASS_NAMES[predicted_labels[i]]}\nActual: {CLASS_NAMES[true_labels[i]]}")
        plt.axis('off')
    plt.tight_layout()
    plt.show()

def main():
    # Load model and data
    model, x_test, y_test = load_assets()
    
    # Make and display predictions
    print("Generating predictions for 3 random images...")
    plot_predictions(model, x_test, y_test)

if __name__ == "__main__":
    main()